lib/navigation/preprocess_imu.cpp

```diff
@@ -1,56 +1,133 @@
+  }
+
+  const core::Float iqr = quartiles.q3 - quartiles.q1;
+  core::Float lower_bound;
+  core::Float upper_bound;
+  if (num_reliable_accelerometers_ == core::kNumImus) {
+    lower_bound = quartiles.median - 1.5 * iqr;
+    upper_bound = quartiles.median + 1.5 * iqr;
+  } else {
+    lower_bound = quartiles.median - 1.2 * iqr;
+    upper_bound = quartiles.median + 1.2 * iqr;
+  }
+
+  for (size_t i = 0; i < core::kNumImus; ++i) {
+    // converts outliers or unreliables to medians, updates number of consecutive outliers for each
+    // sensor
+    if (are_imus_reliable_.at(i) == false) {
+      accelerometer_data.at(i) = quartiles.median;
+    } else if (imu_data.at(i) < lower_bound || imu_data.at(i) > upper_bound) {
+      accelerometer_data.at(i) = quartiles.median;
+      ++num_outliers_per_imu_.at(i);
+    } else {
+      num_outliers_per_imu_.at(i) = 0;
+    }
+  }
+  return accelerometer_data;
 }
 
-void ImuPreprocessor::checkReliable(const core::ImuData &imu_data)
+template<std::size_t N>
+Quartiles ImuPreprocessor::getQuartiles(const std::array<core::Float, N> &clean_accelerometer_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - check how many times an individual imu
-  has been an outlier in a row (outlier imus)
-  -if outlier_imus_[i] > n (tbd), mark imu as unreliable
-  in reliable imus.
-  -if number unreliable in reliable_imus > 1, fail state
-
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
+  std::array<core::Float, N> accelerometer_data_copy;
+  std::copy(clean_accelerometer_data.begin(),
+            clean_accelerometer_data.end(),
+            accelerometer_data_copy.begin());
+
+  std::sort(accelerometer_data_copy.begin(), accelerometer_data_copy.end());
+
+  Quartiles quartiles;
+  quartiles.q1     = getSpecificQuartile(accelerometer_data_copy, 0.25);
+  quartiles.median = getSpecificQuartile(accelerometer_data_copy, 0.5);
+  quartiles.q3     = getSpecificQuartile(accelerometer_data_copy, 0.75);
+  return quartiles;
+}
+
+template<std::size_t N>
+core::Float ImuPreprocessor::getSpecificQuartile(
+  const std::array<core::Float, N> &clean_accelerometer_data, core::Float quartile)
+{
+  const core::Float index_quartile       = (num_reliable_accelerometers_ - 1) * quartile;
+  const std::uint8_t index_quartile_high = static_cast<int>(std::ceil(index_quartile));
+  const std::uint8_t index_quartile_low  = static_cast<int>(std::floor(index_quartile));
+  const core::Float quartile_value       = (clean_accelerometer_data.at(index_quartile_high)
+                                      + clean_accelerometer_data.at(index_quartile_low))
+                                     / 2.0;
+  return quartile_value;
+}
+
+SensorChecks ImuPreprocessor::checkReliable()
+{  // changes reliable sensor to false if max consecutive outliers are reached
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    if (num_outliers_per_imu_.at(i) >= kNumAllowedImuFailures_) {
```

Comment:
If one IMU fails, since you don't reset anything and keep checking against this condition would you not keep reducing the number of reliable imus. Maybe put in a check here for if you've already classified the imu as faulty?